---
title: "Compare VCF files"
author: "Dave Tang"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(cowplot)
library(UpSetR)
theme_set(theme_bw())
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = ".")
```

VCF files are really just tab-delimited files (with additional metadata); we can use `read_tsv` to load VCF files into R. In addition we will add variant IDs (`vid`) using the `chrom`, `id`, `ref`, and `alt` columns in each table and add a variant type column (`type`).

```{r load_vcf, message=FALSE, warning=FALSE}
read_vcf <- function(file){
  vcf_colnames <- c("chrom", "pos", "id", "ref", "alt", "qual", "filter", "info", "format", "sample")
  read_tsv(file, col_names = vcf_colnames, comment = "#") %>%
    mutate(type = case_when(nchar(ref) == 1 & nchar(alt) == 1 ~ "snv",
                            nchar(ref) > 1 & nchar(alt) == 1 ~ "del",
                            nchar(ref) == 1 & nchar(alt) > 1 ~ "ins",
                            TRUE ~ "complex")) %>%
    mutate(vid = paste(chrom, pos, ref, alt, sep = "_")) %>%
    select(vid, everything())
}
fb_variant <- read_vcf(file = "../benchmark/test_31.fb.vcf.gz")
bt_variant <- read_vcf(file = "../benchmark/test_31.bt.vcf.gz")
hc_variant <- read_vcf(file = "../benchmark/test_31.hc.vcf.gz")
```

Check out the first few lines.

```{r check_vcf_ns}
head(hc_variant)
```

An UpSet plot can nicely show the number of common variants.

```{r upsetr_variant_list}
variant_list <- list(BCFtools = bt_variant$vid,
                     FreeBayes = fb_variant$vid,
                     HaplotypeCaller = hc_variant$vid)

upset(fromList(variant_list),
      order.by = "freq",
      point.size = 5,
      text.scale = 1.5)
```

We can see that 92.4% of the variants are commonly called by all three tools. FreeBayes has the largest number of private variants.

Compare SNVs.

```{r compare_snvs}
snvs <- list(BCFtools = bt_variant %>% filter(type == "snv") %>% pull(vid),
             FreeBayes = fb_variant %>% filter(type == "snv") %>% pull(vid),
             HaplotypeCaller = hc_variant %>% filter(type == "snv") %>% pull(vid))

upset(fromList(snvs),
      order.by = "freq",
      point.size = 5,
      text.scale = 1.5)
```

Check the mutation log to see if those private SNVs are "real" or not.

```{r load_mutation_log, message=FALSE, warning=FALSE}
mut_log <- read_tsv(file = "../benchmark/test_31_mutation.log", col_names = c("pos", "ref", "alt"))

hc_variant %>% filter(type == "snv") %>% pull(vid) -> hc_snv
setdiff(hc_snv, union(bt_variant$vid, fb_variant$vid)) %>%
  as.data.frame() %>%
  separate(col = ".", sep = "_", into = c("chrom", "pos", "ref", "alt")) %>%
  mutate(pos = as.numeric(pos)) %>%
  full_join(y = mut_log, by = "pos") %>%
  filter(!is.na(ref.x)) %>%
  mutate(real = case_when(ref.x == ref.y & alt.x == alt.y ~ TRUE,
                          TRUE ~ FALSE)) %>%
  summarise(hc_real_percent = sum(real)*100/n())

bt_variant %>% filter(type == "snv") %>% pull(vid) -> bt_snv
setdiff(bt_snv, union(hc_variant$vid, fb_variant$vid)) %>%
  as.data.frame() %>%
  separate(col = ".", sep = "_", into = c("chrom", "pos", "ref", "alt")) %>%
  mutate(pos = as.numeric(pos)) %>%
  full_join(y = mut_log, by = "pos") %>%
  filter(!is.na(ref.x)) %>%
  mutate(real = case_when(ref.x == ref.y & alt.x == alt.y ~ TRUE,
                          TRUE ~ FALSE)) %>%
  summarise(bt_real_percent = sum(real)*100/n())
```

Compare insertions.

```{r compare_ins}
insertions <- list(BCFtools = bt_variant %>% filter(type == "ins") %>% pull(vid),
                   FreeBayes = fb_variant %>% filter(type == "ins") %>% pull(vid),
                   HaplotypeCaller = hc_variant %>% filter(type == "ins") %>% pull(vid))

upset(fromList(insertions),
      order.by = "freq",
      point.size = 5,
      text.scale = 1.5)
```

Compare deletions.

```{r compare_del}
deletions <- list(BCFtools = bt_variant %>% filter(type == "del") %>% pull(vid),
                  FreeBayes = fb_variant %>% filter(type == "del") %>% pull(vid),
                  HaplotypeCaller = hc_variant %>% filter(type == "del") %>% pull(vid))

upset(fromList(deletions),
      order.by = "freq",
      point.size = 5,
      text.scale = 1.5)
```

## Complex UpSet plot

Visualise using a complex UpSet plot. First we need to prepare the data frame.

```{r variant_df}
rbind(bt_variant %>% select(vid, type),
      hc_variant %>% select(vid, type),
      fb_variant %>% select(vid, type)) %>%
  distinct(vid, type) -> all_variant

all_variant$bt <- all_variant$vid %in% bt_variant$vid
all_variant$hc <- all_variant$vid %in% hc_variant$vid
all_variant$fb <- all_variant$vid %in% fb_variant$vid
```

UpSet plot with variant type.

```{r complex_upset, fig.width=16, fig.height=6}
ComplexUpset::upset(all_variant,
                    c("bt", "hc", "fb"),
                    base_annotations = list(
                      'Intersection size' = ComplexUpset::intersection_size(
                        counts = FALSE,
                        aes = aes(fill = type)
                      )
                    ),
                    width_ratio = 0.1
)
```

UpSet plot with variant type as another plot.

```{r complex_upset_2, fig.width=16, fig.height=6}
ComplexUpset::upset(all_variant,
                    c("bt", "hc", "fb"),
                    annotations = list(
                      'Type' = list(
                        aes = aes(x = intersection, fill = type),
                        geom = list(
                          geom_bar(stat = 'count', position = 'fill'),
                          scale_y_continuous(labels = scales::percent_format())
                        )
                      )
                    ),
                    width_ratio = 0.1
)
```